File: sdk/python/src/plinth/workers.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 The Plinth Authors
"""Worker registration client for the durable workflow executor (v0.5).

Mirrors the workspace service's ``/v1/workers`` endpoints. The client is
attached to the :class:`Plinth` facade as ``client.workers`` and is
typically used only by the worker process and ops tooling — application
code rarely registers a worker directly.
"""

from __future__ import annotations

import os
import socket
from typing import TYPE_CHECKING

from .exceptions import WorkerNotFound
from .models import Worker

if TYPE_CHECKING:
    from .client import Plinth
    from ._http import HTTPClient


class WorkersClient:
    """Workspace-service ``/v1/workers`` client.

    The methods talk to the *workspace* service (where worker rows live)
    even though they're attached to the gateway-shaped :class:`Plinth`
    facade. Workers are workspace-scoped state, not gateway-scoped.
    """

    def __init__(self, http: HTTPClient) -> None:
        self._http = http

    def register(
        self,
        *,
        hostname: str | None = None,
        pid: int | None = None,
    ) -> Worker:
        """Register a new worker process.

        ``hostname`` and ``pid`` default to the current process's values
        so a typical worker just calls ``client.workers.register()``.
        """

        payload: dict = {
            "hostname": hostname if hostname is not None else socket.gethostname(),
            "pid": pid if pid is not None else os.getpid(),
        }
        response = self._http.post("/v1/workers/register", json=payload)
        return Worker.model_validate(response.json())

    def heartbeat(self, worker_id: str) -> Worker:
        """Bump ``last_heartbeat_at`` for ``worker_id``."""
        response = self._http.post(
            f"/v1/workers/{worker_id}/heartbeat",
            not_found_class=WorkerNotFound,
        )
        return Worker.model_validate(response.json())

    def drain(self, worker_id: str) -> Worker:
        """Mark ``worker_id`` as ``draining`` (graceful shutdown signal)."""
        response = self._http.post(
            f"/v1/workers/{worker_id}/drain",
            not_found_class=WorkerNotFound,
        )
        return Worker.model_validate(response.json())

    def list(self, *, status: str | None = None) -> list[Worker]:
        """List registered workers (optionally filtered by ``status``)."""
        params = {"status": status} if status else None
        data = self._http.get_json("/v1/workers", params=params)
        return [Worker.model_validate(w) for w in data.get("workers", [])]


__all__ = ["WorkersClient"]

File: sdk/go/plinth/workers.go
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 The Plinth Authors

package plinth

import (
	"context"
	"net/url"
	"os"
)

// WorkersClient is the workspace service's worker registration
// surface (v0.5). Used by the workflow worker harness — application
// code rarely registers a worker directly.
type WorkersClient struct {
	http *HTTPClient
}

// NewWorkersClient constructs a client bound to a workspace HTTPClient.
// Exported so tests can wire one up directly.
func NewWorkersClient(workspaceHTTP *HTTPClient) *WorkersClient {
	return &WorkersClient{http: workspaceHTTP}
}

// Register registers a new worker process. Hostname and PID default
// to the current process's values when reg.Hostname / reg.PID are nil.
func (c *WorkersClient) Register(ctx context.Context, reg WorkerRegistration) (*Worker, error) {
	body := map[string]any{}
	if reg.Hostname != nil {
		body["hostname"] = *reg.Hostname
	} else {
		body["hostname"] = safeHostname()
	}
	if reg.PID != nil {
		body["pid"] = *reg.PID
	} else {
		body["pid"] = os.Getpid()
	}
	var w Worker
	if err := c.http.PostJSON(ctx, "/v1/workers/register", &w, WithJSON(body)); err != nil {
		return nil, err
	}
	return &w, nil
}

// Heartbeat bumps last_heartbeat_at for workerID.
func (c *WorkersClient) Heartbeat(ctx context.Context, workerID string) (*Worker, error) {
	var w Worker
	err := c.http.PostJSON(
		ctx,
		"/v1/workers/"+EncodePathSegment(workerID)+"/heartbeat",
		&w,
		WithNotFoundCode(ErrWorkerNotFound.Code),
	)
	if err != nil {
		return nil, err
	}
	return &w, nil
}

// Drain marks workerID as draining (graceful-shutdown signal).
func (c *WorkersClient) Drain(ctx context.Context, workerID string) (*Worker, error) {
	var w Worker
	err := c.http.PostJSON(
		ctx,
		"/v1/workers/"+EncodePathSegment(workerID)+"/drain",
		&w,
		WithNotFoundCode(ErrWorkerNotFound.Code),
	)
	if err != nil {
		return nil, err
	}
	return &w, nil
}

// List returns registered workers, optionally filtered by status.
// Pass "" for status to list every worker.
func (c *WorkersClient) List(ctx context.Context, status string) ([]Worker, error) {
	q := url.Values{}
	if status != "" {
		q.Set("status", status)
	}
	var resp struct {
		Workers []Worker `json:"workers"`
	}
	if err := c.http.GetJSON(ctx, "/v1/workers", &resp, WithQuery(q)); err != nil {
		return nil, err
	}
	return resp.Workers, nil
}

// safeHostname returns os.Hostname() or "" on failure. Mirrors the
// TS SDK's safeHostname helper — the workspace service tolerates an
// empty string so we degrade rather than fail boot.
func safeHostname() string {
	h, err := os.Hostname()
	if err != nil {
		return ""
	}
	return h
}
